src/sshagent/DatabaseAgent.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ *  Copyright (C) 2017 Toni Spets <toni.spets@iki.fi>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "DatabaseAgent.h"
+#include "sshagent/KeeAgentSettings.h"
+
+DatabaseAgent::DatabaseAgent(DatabaseWidget *parent) : QObject(parent), m_widget(parent)
+{
+    connect(parent, SIGNAL(currentModeChanged(DatabaseWidget::Mode)), this, SLOT(databaseModeChanged(DatabaseWidget::Mode)));
+}
+
+DatabaseAgent::~DatabaseAgent()
+{
+    emit databaseModeChanged(DatabaseWidget::LockedMode);
+}
+
+void DatabaseAgent::databaseModeChanged(DatabaseWidget::Mode mode)
+{
+    auto keys = entryKeys(m_widget->database());
+
+    if (mode == DatabaseWidget::LockedMode && m_sentKeys) {
+        for (QSharedPointer<OpenSSHKey> e : keys) {
+            AgentClient::instance()->removeIdentity(*e.data());
+        }
+
+        m_sentKeys = false;
+    }
+
+    if ((mode == DatabaseWidget::ViewMode
+            || mode == DatabaseWidget::EditMode)
+            && !m_sentKeys) {
+        for (QSharedPointer<OpenSSHKey> e : keys) {
+            AgentClient::instance()->addIdentity(*e.data());
```

Comment:
Here `lifetime` and `confirm` should be passed, too.

Reply:
Need to do a bit of refactoring there, it was written before constraints existed.